metrics/cloudwatch2/cloudwatch2.go

```diff
@@ -0,0 +1,239 @@
+}
+
+// NewHistogram returns a histogram. Under the covers, there is no distinctions
+// in CloudWatch for how Counters/Histograms/Gauges are reported, so this
+// just wraps a cloudwatch2.Counter.
+func (cw *CloudWatch) NewHistogram(name string) metrics.Histogram {
+	return convert.NewCounterAsHistogram(cw.NewCounter(name))
+}
+
+// WriteLoop is a helper method that invokes Send every time the passed
+// channel fires. This method blocks until the channel is closed, so clients
+// probably want to run it in its own goroutine. For typical usage, create a
+// time.Ticker and pass its C channel to this method.
+func (cw *CloudWatch) WriteLoop(c <-chan time.Time) {
+	for range c {
+		if err := cw.Send(); err != nil {
+			cw.logger.Log("during", "Send", "err", err)
+		}
+	}
+}
+
+// Send will fire an API request to CloudWatch with the latest stats for
+// all metrics. It is preferred that the WriteLoop method is used.
+func (cw *CloudWatch) Send() error {
+	cw.mtx.RLock()
+	defer cw.mtx.RUnlock()
+	now := time.Now()
+
+	var datums []cloudwatch.MetricDatum
+
+	cw.counters.Reset().Walk(func(name string, lvs lv.LabelValues, values []float64) bool {
+		datums = append(datums, cloudwatch.MetricDatum{
+			MetricName:      aws.String(name),
+			Dimensions:      makeDimensions(lvs...),
+			StatisticValues: stats(values),
+			Timestamp:       aws.Time(now),
+		})
+		return true
+	})
+
+	var batches [][]cloudwatch.MetricDatum
+	for len(datums) > 0 {
+		var batch []cloudwatch.MetricDatum
+		lim := len(datums)
+		if lim > maxConcurrentRequests {
+			lim = maxConcurrentRequests
+		}
+		batch, datums = datums[:lim], datums[lim:]
+		batches = append(batches, batch)
+	}
+
+	var g errgroup.Group
+	for _, batch := range batches {
+		batch := batch
+		g.Go(func() error {
+			cw.sem <- struct{}{}
+			defer func() {
+				<-cw.sem
+			}()
+			req := cw.svc.PutMetricDataRequest(&cloudwatch.PutMetricDataInput{
+				Namespace:  aws.String(cw.namespace),
+				MetricData: batch,
+			})
+			_, err := req.Send()
+			return err
+		})
+	}
+	return g.Wait()
+}
+
+var zero = float64(0.0)
+
+// Just build this once to reduce construction costs whenever
+// someone does a Send with no aggregated values.
+var zeros = cloudwatch.StatisticSet{
+	Maximum:     &zero,
+	Minimum:     &zero,
+	Sum:         &zero,
+	SampleCount: &zero,
+}
+
+func stats(a []float64) *cloudwatch.StatisticSet {
+	count := float64(len(a))
+	if count == 0 {
+		return &zeros
+	}
+
+	var sum float64
+	var min = math.MaxFloat64
+	var max = math.MaxFloat64 * -1
+	for _, f := range a {
+		sum += f
+		if f < min {
+			min = f
+		}
+		if f > max {
+			max = f
+		}
+	}
+
+	return &cloudwatch.StatisticSet{
+		Maximum:     &max,
+		Minimum:     &min,
+		Sum:         &sum,
+		SampleCount: &count,
+	}
+}
+
+func makeDimensions(labelValues ...string) []cloudwatch.Dimension {
+	dimensions := make([]cloudwatch.Dimension, len(labelValues)/2)
+	for i, j := 0, 0; i < len(labelValues); i, j = i+2, j+1 {
```

Comment:
Ha! Wild. Nice.

Reply:
I can't take credit for this... I lifted this directly from the existing `cloudwatch` package.